packages/daimo-api/src/contract/foreignCoinIndexer.ts

```diff
@@ -0,0 +1,209 @@
+import { ProposedSwap, guessTimestampFromNum } from "@daimo/common";
+import { Pool } from "pg";
+import { Address, Hex, bytesToHex, getAddress } from "viem";
+
+import { Transfer } from "./homeCoinIndexer";
+import { NameRegistry } from "./nameRegistry";
+import { chainConfig } from "../env";
+import { UniswapClient } from "../network/uniswapClient";
+import { ViemClient } from "../network/viemClient";
+import { ForeignToken, fetchTokenList } from "../server/coinList";
+import { addrTxHashKey } from "../utils/indexing";
+import { retryBackoff } from "../utils/retryBackoff";
+
+// An in/outbound swap coin transfer with swap coin metadata.
+export type ForeignTokenTransfer = Transfer & {
+  foreignToken: ForeignToken;
+};
+
+/* Tracks non-USDC coin transfers.
+ * The lifecycle of a foreign coin is a bit complex:
+ * - An external user sends a foreign token to Daimo address.
+ * - Daimo address treats the inbound transfer as a "pending swap".
+ * - Daimo address sends a "swap" transaction to Uniswap / DEX, generating a
+ *   outbound swap transfer log.
+ * - The DEX generates an inbound transfer of Home coin to Daimo address.
+ * - Foreign Coin Indexer helps map the final inbound Home coin transfer to the
+ *   original inbound foreign token transfer.
+ */
+export class ForeignCoinIndexer {
+  private foreignTokens: Map<Address, ForeignToken> = new Map();
+  private allTransfers: ForeignTokenTransfer[] = [];
+
+  private pendingSwapsByAddr: Map<Address, ForeignTokenTransfer[]> = new Map(); // inbound transfers without a corresponding outbound swap
+  private sendsByAddrTxHash: Map<string, ForeignTokenTransfer> = new Map(); // outbound transfers
+  private correspondingReceiveOfSend: Map<string, ForeignTokenTransfer> =
+    new Map(); // inbound foreign token transfer corresponding the outbound swap send.
+
+  private listeners: ((transfers: ForeignTokenTransfer[]) => void)[] = [];
+
+  constructor(
+    private nameReg: NameRegistry,
+    private vc: ViemClient,
+    private uc: UniswapClient
+  ) {}
+
+  async preIndexingInit(pg: Pool) {
+    this.foreignTokens = await fetchTokenList();
+    await retryBackoff(`createFilteredTransfersView`, async () => {
+      await pg.query(`CREATE MATERIALIZED VIEW IF NOT EXISTS filtered_erc20_transfers as (SELECT
+        et.*
+      FROM
+        erc20_transfers et
+        JOIN names n ON n.addr = et.f
+          OR n.addr = et.t);`);
+    });
```

Comment:
hmm i worry about having db migrations distributed around the codebase

recommend putting with the others in db.ts

```sql
CREATE MATERIALIZED VIEW IF NOT EXISTS filtered_erc20_transfers
AS SELECT et.*
FROM erc20_transfers et
JOIN names n ON n.addr = et.f OR n.addr = et.t
```

Reply:
moved to shovel/watcher init